### ISSUE TITLE
First time experience

### ISSUE THREAD

**Issue by MarcusKlerks** [jorijnsmit/onlineChecker#1]
Hi there,
I saw your change to the wiki for presence detection running as a service. I found it interesting enough to give it a try. I finally got it running and I am happy with it but I have some remarks. 
This is what I did:
Since the Python script is compatible with Python 3.6 and upwards I had to install a higher version than Python 3.5 that is standard on Raspberry (Stretch). I installed version 3.7.3 using the instructions at https://gist.github.com/SeppPenner/6a5a30ebc8f79936fa136c524417761d.

That installed Python 3.7.3 in /usr/local/bin so I had to make an adaption to the service (next to change the working directory and user):
ExecStart=/usr/local/bin/python3.7 -u onlineChecker.py xxx.xxx.x.xxx 10 5 100

After starting the service, I got an error due to a missing module (requests). After installing this the script ran fine.
Two remarks: 
Detection of my phone works better with arping, so I changed pinging into:
pingReply = subprocess.call('sudo arping -q -c1 -W 1 '+ DEVICE + ' > /dev/null', shell=True)
Could this be an option?

I am used to protect the switches used to indicate presence with a passcode. I have added that to credentials.py:
PASS_CODE = 'same password as in Domoticz settings for Light/Switch Protection'
And changed the code in domoCommand(cmd) into:
print(domoRequest(f'type=command&param=switchlight&idx={IDX}&switchcmd={cmd}&passcode={PASS_CODE}')['status'])

After applying the changes it runs fine for me. 
Thanks, Mark

BTW. I think your effort deserves more advertisement than just the short line in the Wiki (who besides me has noticed that).

**Comment by jorijnsmit** [comment]
Thanks so much for your feedback! 
- In regards to the Python version: I did not realise 3.5 is standard on the RPi. It will probably make it a lot easier for everybody if it runs on 3.5 (and thus on `/usr/local/python3`). The only reason I chose 3.6 is so that I would be able to use f-strings but I think not using those is overcomeable.
- I'll add instructions for installing `requests` and `arping`:
- The original script used `arping` as well. I'll do some testing as to see which one works better, or maybe add a switch to allow choosing between both.
- I'll add support for PASS_CODE.

I had fun making this and will do my best to improve it. Maybe it will be able to replace the old script on the wiki eventually!

**Comment by MarcusKlerks** [comment]
Making it compatible with Python 3.5 is a great idea and I think it will be welcomed by a lot of users. The other items are also welcome. I think the Readme should be more explanatory for less experienced users. I'll be available for future review.

**Comment by jorijnsmit** [comment]
Runs on python3.5 now, requests is mentioned in the readme and protected switches are supported. Using `arping` (or other protocols) instead of `ping` will remain an issue for later.

I might also include Chopper_Rob's old script from the Domoticz wiki page in this repo for legacy reasons. This will allow me to update the wiki page and explain presence detection using this script instead.

**Comment by MarcusKlerks** [comment]
I have reviewed your changes (not tested yet) and it leaves me with some remarks/questions:


  *   On installing the requests library: I had to execute ‘sudo pip3 install requests’ (Raspberry PI3, Stretch, Python 2.7, Python 3.5).


  *   I see that you changed credentials.py to hold the device and ping characteristics where they formerly where part of the .service file. Although this seems handy at first sight, the downside is that the use now is limited to check only one device. In  my setup I created different .service files for the phones of my wife and myself. That worked just fine (I made my own few changes to make it run with Python 3.5).

So question: why this change?


  *   For less experienced users I think the installation of .service file(s) needs some more detailed instructions:
- copy the adapted .service file into /etc/systemd/system/
- sudo systemctl daemon-reload          # reloads services, must be done after every change
- sudo systemctl enable <name>.service    # enables the service to start at boot, use your own naming for <name>
- sudo systemctl start <name>.service        # start the service


  *   In the mean time I changed phones from Windows Phone to Android. Where the Windows Phone worked better with arping, the Android reacts better on ping. But in both cases I see that the sys.log is ‘flooded’ with disconnect/connect logs. Do you have the same experience?


  *   Kudos for the changes in onlineChecker.py, makes it better readable.

Regards, Mark

Sent from Windows

________________________________

**Comment by jorijnsmit** [comment]
Concerning the use of `pip` and `systemctl`: I don't want to spend too much time on this here. I think their use is already documented sufficiently elsewhere.

In regards to your issue of tracking multiple devices: the reason I got rid of the parameters in the .service file is that I think it is the wrong place to be tinkering with them. Like you mention, `systemctl` has to re-read the .service file. On top of that editing the file needs admin rights and it is more difficult to backup. By putting everything in a credentials file these problems are solved. I have added an issue (#3) to support multiple credentials files and thus multiple devices.

The 'disconnects' you mention in my experience are just the Wi-Fi going into sleep mode on the device. Looking into your battery saving/sleep/networking settings on your devices could solve (some). An increase in the cooldown period might help also. 

I ran successful tests where I replace `ping` with an API call to my router which checks whether the device is still connected. This is a more reliable method in my opinion but is very dependent on whether your router has an API available.
